lib/cstruct.ml

```diff
@@ -370,6 +370,35 @@ let of_bytes ?allocator buf =
     blit_from_bytes buf 0 c 0 buflen;
     set_len c buflen
 
+let string_fold ~f ~z str =
+  let st = ref z in
+  ( String.iter (fun c -> st := f !st c) str  ; !st )
+
+let of_hex str =
+  let hexdigit = function
+    | 'a' .. 'f' as x -> int_of_char x - 87
+    | 'A' .. 'F' as x -> int_of_char x - 55
+    | '0' .. '9' as x -> int_of_char x - 48
+    | x               -> Format.ksprintf invalid_arg "of_hex: `%c'" x
+  in
+  let whitespace = function
+    | ' ' | '\t' | '\r' | '\n' -> true
+    | _                        -> false
+  in
+  match
+    string_fold
+      ~f:(fun (cs, i, acc) -> function
+          | char when whitespace char -> (cs, i, acc)
+          | char ->
+            match (acc, hexdigit char) with
```

Comment:
I think printing the offset where the invalid byte was spotted (if it is invalid) would also be a good addition for debugging. For example by passing `i` to `hexdigit` here.

Reply:
see above, it is now printed